Mockable API

markus-wa opened the pull request:
pinging @micvbang 

This is my counter proposal to #57.

As we've discussed there I looked into interface generation to avoid godoc and Intellisense issues.
I found that another nice side-effect of this approach is that the impact on consumers is a bit lighter as they won't need to change their variable/field types from `*Parser` to `Parser` now (though for `GameState` and `Participants` that 'issue' remains).

I also added a new package `mock` that provides basic but (hopefully) extendable mocks for the various interfaces, so not everyone needs to re-implement them unless there's a specific need.

Lastly I added an example on how to use the mocks to write unit tests: https://github.com/markus-wa/demoinfocs-golang/tree/mockable-api/examples/mocking

For the last two I took a lot of inspiration from your gist which was very helpful 😄.

It would be great if you could check if there's something I missed or if this would work for you. Thanks in advance.

markus-wa commented on mock/parser.go:
```diff
@@ -0,0 +1,167 @@
+
+// RegisterNetMessageHandler is a mock-implementation of IParser.RegisterNetMessageHandler().
+func (p *Parser) RegisterNetMessageHandler(handler interface{}) dp.HandlerIdentifier {
+	return dp.HandlerIdentifier(p.msgDispatcher.RegisterHandler(handler))
```
good bot

micvbang commented:
I'm beginning to look through it now, and just an initial thought; it does not seem like I can avoid changing the type signature in my own code using the library. I still have to change it to `dem.IParser`.

This is the case because I want to my tests to test my full pipeline, including my own wrapper of the demoinfocs-golang `Parser`. In order to be able to control the behavior of `Parser`, I have to use the interface type so that I can give it the new `fake` implementation.

This is not a bad thing at all, just thought I'd note it :slightly_smiling_face:

micvbang commented:
I think it would be really helpful to have a primitive that allows the user to not care about the id of events per default, but allow her to do it if it really matters for the test.

In my own mock implementation I did this by keeping track of the current event id, something like this:

```go
func (p *parserMock) addEventDispatchFrame(frame int, event interface{}) {
	if frame > p.frameFinal {
		p.frameFinal = frame
	}

	p.eventDispatchQueue[frame] = append(p.eventDispatchQueue[frame], event)
	p.addFrameWithDispatch(frame)
}

func (p *parserMock) addEventDispatch(event interface{}) int {
	p.addEventDispatchFrameCurrent++
	p.addEventDispatchFrame(p.addEventDispatchFrameCurrent, event)
	return p.addEventDispatchFrameCurrent
}
```

Here, the user can let the mock track the current frame by using `addEventDispatch`, and in cases where a certain event should happen in the same frame (perhaps multiple events in the same frame), she can use `addEventDispatchFrame`. Both of them return the frame that was added, so it's easy to mix usage of the two

micvbang commented:
I'll also add that the most important part of this PR, to me, is the addition of the interfaces. With them, we can implement any kind of API that we want for our mocking, and everyone doesn't really have to agree about its implementation  :slightly_smiling_face:

markus-wa commented:
Thanks for the feedback.

I actually intentionally didn't add helpers as they are easy to add if needed, but maybe it's a good idea - might look into it again.

What I really didn't consider was multiple events in the same frame, we should certainly cover that, maybe with `map[int][]interface{}`.

And yeah the mocks in `fake` are just there as utility, if needed it's always possible to write your own.

markus-wa commented:
I fixed up the issue with multiple events on the same frame and I've also added `MockEvents(...interface{})` and `MockEventsFrame(int, ...interface{})` as per your recommendation. :slightly_smiling_face:

markus-wa commented:
Shall we merge this? @micvbang

micvbang commented:
I believe so!

markus-wa commented:
Done
